### PR TITLE
tests: Use kata-deploy scripts to build the artefacts as part of our CI

### PR DESCRIPTION
**Pull request by fidencio**
We've done this as part of the CCv0 branch, and now we're starting to slowly move the work to the `main` branch.

The main motivations for doing this are:
* Ensure we test our artefacts in the exactly same way they'll be built as part of the release process
* Decrease the complexity of the CI

### CODE REVIEW

**Comment by fidencio** [comment]
Right now we have the following related failures:
* [ ] Fedora builds are failing as `make $component-tarball` doesn't seem to be fully working with podman, I've started a discussion [here](https://katacontainers.slack.com/archives/C879ACQ00/p1674806962031269).
* [ ] Dragonball build is failing as we need to add to kata-deploy the ability to build the dragonball-kernel, which has UPS support, [I shared a branch](https://github.com/fidencio/kata-containers/commits/topic/upcall-follow-up) with @studychao some time ago that does that.

Apart from this:
* [ ] ARM failures are not related

**Comment by fidencio** [comment]
@jongwu, mind to give me some help to debug what's going on with http://jenkins.katacontainers.io/job/kata-containers-2.0-tests-ubuntu-ARM-PR/2680/console ?

**Comment by fidencio** [comment]
@BbolroC, mind to give me some help to debug what's going on with http://jenkins.katacontainers.io/job/kata-containers-2.0-tests-ubuntu-s390x-main-PR/1431/console ?

**Comment by BbolroC** [comment]
> @BbolroC, mind to give me some help to debug what's going on with http://jenkins.katacontainers.io/job/kata-containers-2.0-tests-ubuntu-s390x-main-PR/1431/console ?

yeah, it is `prost-build` error. We need to get https://github.com/kata-containers/kata-containers/pull/6245 ported to `main.

**Comment by BbolroC** [comment]
I've created a PR for `protobuf`: https://github.com/kata-containers/kata-containers/pull/6273

**Comment by fidencio** [comment]
> I've created a PR for `protobuf`: [kata-containers/kata-containers#6273](https://github.com/kata-containers/kata-containers/pull/6273)

Is that already merged on main?  If so, I don't get why it's still failing here (this PR is targeting main).

**Comment by BbolroC** [comment]
> > I've created a PR for `protobuf`: [kata-containers/kata-containers#6273](https://github.com/kata-containers/kata-containers/pull/6273)
> 
> Is that already merged on main? If so, I don't get why it's still failing here (this PR is targeting main).

No, the merged one is for CCv0. The newly created one is what I've just cherry-picked a few minutes ago.

**Comment by BbolroC** [comment]
https://github.com/kata-containers/kata-containers/pull/6143 has been merged, which will set you free from the CI failures for s390x and ppc64le. :wink:

**Comment by BbolroC** [comment]
regarding the s390x-main failure:

```
sed: can't read /home/jenkins/workspace/kata-containers-2.0-tests-ubuntu-s390x-main-PR/
go/src/github.com/kata-containers/kata-containers/tools/packaging/kata-deploy/local-build/
build/shim-v2/destdir//opt/kata/share/defaults/kata-containers/configuration-fc.toml: No such file or directory

```

It seems the following line should run conditionally:

https://github.com/kata-containers/kata-containers/blob/0d2a7f8324bc1fcd7d9b11b9f957c67968afc302/tools/packaging/static-build/shim-v2/build.sh#L51

**Comment by Amulyam24** [comment]
Hi @fidencio, regarding the failure on ppc64le : 
```
 ---> Running in e1ed11cf1a7c
Removing intermediate container e1ed11cf1a7c
 ---> 670234ff6401
Successfully built 670234ff6401
Successfully tagged shim-v2-builder:latest
../../utils.mk:151: "WARNING: powerpc64le-unknown-linux-musl target is unavailable"
Makefile:54: *** "ERROR: invalid architecture: 'powerpc64le'".  Stop.
make[1]: *** [tools/packaging/kata-deploy/local-build/Makefile:40: shim-v2-tarball-build] Error 2
make[1]: Leaving directory '/home/jenkins/workspace/kata-containers-2.0-tests-ppc64le-containerd-k8s-ubuntu-20-04-PR/go/src/github.com/kata-containers/kata-containers'
make: *** [tools/packaging/kata-deploy/local-build/Makefile:70: shim-v2-tarball] Error 2
[install_runtime.sh:175] ERROR: sudo -E PATH=/home/jenkins/workspace/kata-containers-2.0-tests-ppc64le-containerd-k8s-ubuntu-20-04-PR/go/bin:/usr/local/go/bin:/usr/sbin:/sbin:/home/jenkins/workspace/kata-containers-2.0-tests-ppc64le-containerd-k8s-ubuntu-20-04-PR/go/bin:/usr/local/go/bin:/usr/sbin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin make shim-v2-tarball
[install_kata.sh:58] ERROR: .ci/install_runtime.sh 
```
I can see that the error is coming from here -- https://github.com/kata-containers/kata-containers/blob/0d2a7f8324bc1fcd7d9b11b9f957c67968afc302/src/runtime-rs/Makefile#L54. Not sure why though, we don't proceed if the `ARCH` is `powerpc64le` right, Is my understanding correct?

**Comment by fidencio** [comment]
> Regarding the s390x-main failure:
> 
> ```
> sed: can't read /home/jenkins/workspace/kata-containers-2.0-tests-ubuntu-s390x-main-PR/
> go/src/github.com/kata-containers/kata-containers/tools/packaging/kata-deploy/local-build/
> build/shim-v2/destdir//opt/kata/share/defaults/kata-containers/configuration-fc.toml: No such file or directory
> ```
> 
> It seems the following line should run conditionally:
> 
> https://github.com/kata-containers/kata-containers/blob/0d2a7f8324bc1fcd7d9b11b9f957c67968afc302/tools/packaging/static-build/shim-v2/build.sh#L51
> 
> Others (unit and snap) have passed.

Yep, you had a patch for that for the `CCv0` branch.  Mind to forward-port that to `main`? :-)

**Comment by fidencio** [comment]
> Yep, you had a patch for that for the `CCv0` branch. Mind to forward-port that to `main`? :-)

I've created: https://github.com/kata-containers/kata-containers/pull/6294

**Comment by fidencio** [comment]
@Amulyam24, can you help me to debug this one? http://jenkins.katacontainers.io/job/kata-containers-2.0-tests-ppc64le-containerd-k8s-ubuntu-20-04-PR/629/console

File: .ci/install_qemu.sh [hunk 1]
```diff
@@ -201,29 +60,27 @@ main() {
...
 
-	# Strip "https://" to clone into $GOPATH
-	gopath_qemu_repo=${QEMU_REPO_URL/https:\/\//}
-	# These variables are used by static-build scripts
-	export qemu_version="${CURRENT_QEMU_VERSION}"
-	export qemu_repo="${QEMU_REPO_URL}"
-	run
+	case ${QEMU_ARCH} in
+		"ppc64le")
+			build_and_install_qemu
```
Comment (Amulyam24) [inline .ci/install_qemu.sh:75]: @fidencio, one question here. With this change, the `qemu-ppc64` binary will be installed in `/usr/bin`. However the default path that is being used in the config `/opt/kata/share/defaults/kata-containers/configuration.toml` is
```
[hypervisor.qemu]
path = “/opt/kata/bin/qemu-system-ppc64”
```
If my understanding is correct, this should also be changed right?

Comment (fidencio) [reply .ci/install_qemu.sh:75]: You're correct.  Ideally arch specific script should rexpect the ${PREFIX} being used.

I will adjust that, thanks for pointing this out!

**Comment by jongwu** [comment]
Hi @fidencio -, Arm CI fails for 

> 03:19:16 kata-runtime: /lib/aarch64-linux-gnu/libc.so.6: version `GLIBC_2.32' not found (required by kata-runtime)
03:19:16 kata-runtime: /lib/aarch64-linux-gnu/libc.so.6: version `GLIBC_2.34' not found (required by kata-runtime)

as you build kata-runtime dynamically in ubuntu:22.04 but the CI machine is ubuntu:20.04 which has lower version of glibc.

**Comment by fidencio** [comment]
@jongwu, can you help to debug what's going on with the ARM side?  I'm pretty sure it's something wrong on my side, but I'd appreciate help on this.

**Comment by jongwu** [comment]
Hi @fidencio -, the issue derived from kata start without pflash in configuration. Maybe you forget to set pflash for arm?

**Comment by jongwu** [comment]
This line
https://github.com/kata-containers/tests/blob/6bfe46de754f4ac1872d31749e65a7d149db4f19/.ci/aarch64/install_rom_aarch64.sh#L128

should be
runtime_config_prefix=("/etc"` "/usr/share/defaults" "/opt/kata/share/defaults")

After adding this change, I think arm ci can pass.

@fidencio , can you try it?

**Comment by jongwu** [comment]
@fidencio -, in other way, you can test again depends on my [PR](https://github.com/kata-containers/kata-containers/pull/6467) to this test. There, I set firmware path for qemu on arm. Based on that PR we can build qemu-tarball for arm64 the same way as x86, that is using  build_static_artifact_and_install instead. I think this is a better way.

**Comment by wainersm** [comment]
@fidencio I am trying to debug why http://jenkins.katacontainers.io/job/tests-fedora-35-x86_64-CRIO_K8S_MINIMAL-PR/807/consoleFull failed.

**Comment by littlejawa** [comment]
> @fidencio I am trying to debug why http://jenkins.katacontainers.io/job/tests-fedora-35-x86_64-CRIO_K8S_MINIMAL-PR/807/consoleFull failed.

@wainersm - I've been looking at it, but I admit I'm not sure what I can get from the logs.
I feel the output we have today is not enough :-/

Our script (run.sh) expects an "e2e.log" file that it doesn't find, so we have no output from the e2e run, and no idea what went wrong in there (or did I miss it?).
The logs from the tarball extraction show there are data we could use though, like an "plugins/e2e/errors/global/error.json" file. But they are not available anymore (right?).

Is it possible to copy the full sonobuoy tarball to the "artifacts" in Jenkins, so that we can work on the full output even after the environment has gone away? I don't know if disk size is a constraint on jenkins though.

I can make a PR in that direction, but I'm not even sure my comments are valid - tell me if it makes sense.

**Comment by wainersm** [comment]
> > I was able to reproduce the sonobuoy but I still have not clue what is going on. It seems that the tests didn't even start.
> 
> They did, take a look at the commit message of the last commit. For some reason they do start, they do pass, but it seems a specific sonobuoy test is failing. I've joined the community's Slack and asked for help there.
>

I looked at the commit message. Very good information.
 
> For now, I really would not like this to block us.

Yeah, let's move forward.

File: .ci/install_runtime.sh [hunk 2]
```diff
@@ -57,17 +57,14 @@ NEW_RUNTIME_CONFIG="${PKGDEFAULTSDIR}/configuration.toml"
 clone_katacontainers_repo
 
 build_install_shim_v2(){
-	pushd "$runtime_src_path"
-	make
-	sudo -E PATH=$PATH make install
-	popd
+	build_static_artifact_and_install "shim-v2"
```
Comment (stevenhorsman) [inline .ci/install_runtime.sh:60]: Just to double check I understand this change (as I suspect the other commits will be along a similar pattern) - `build_static_artifact_and_install "shim-v2"` will trigger a bunch of scripts that will eventually get to [`kata-deploy-binaries.sh`](https://github.com/kata-containers/kata-containers/blob/main/tools/packaging/kata-deploy/local-build/kata-deploy-binaries.sh) that will run `install_shimv2` and build the runtime binaries to `kata-deploy-binaries.sh:prefix` which is set to `readonly prefix="/opt/kata"` there and so we can just link to those built files here?

Comment (fidencio) [reply .ci/install_runtime.sh:60]: That's exactly the case.

And the only binary we need to link (or add to the containerd path) is the `containerd-shim-kata-v2`.  All other binaries do not matter, as we use the full path in the specific `configuration.toml`, which is already using `/opt/kata/...`

File: .ci/install_kata_kernel.sh [hunk 3]
```diff
@@ -174,23 +68,16 @@ main() {
...
+			build_and_install_kernel "kernel"
 			;;
 		dragonball)
-			install_kernel "${kernel_type}" $(get_version "assets.kernel-dragonball-experimental.version")
+			build_and_install_kernel "kernel-dragonball-experimental"
 			;;
 		*)
-			info "kernel type '${kernel_type}' not supported"
-			usage 1
+			die_unsupported_kernel_type "$kernel_type"
```
Comment (littlejawa) [inline .ci/install_kata_kernel.sh:80]: I don't see where this call is defined... Am I missing something?
All "die()" calls are in the form "die [message]", is it what you intended?

File: .ci/install_qemu.sh [hunk 4]
```diff
@@ -201,29 +78,38 @@ main() {
...
-		tdx)
-			CURRENT_QEMU_VERSION=$(get_version "assets.hypervisor.qemu.tdx.tag")
-			QEMU_REPO_URL=$(get_version "assets.hypervisor.qemu.tdx.url")
-			qemu_latest_build_url="${jenkins_url}/job/kata-containers-2.0-qemu-tdx-$(uname -m)/${cached_artifacts_path}"
-			;;
 		vanilla)
+			qemu_type="qemu"
+			;;
+		*)
+			die_unsupported_qemu_type "$qemu_type"
```
Comment (littlejawa) [inline .ci/install_qemu.sh:87]: Same comment here: I don't see this call defined or used anywhere else.

Comment (fidencio) [reply .ci/install_qemu.sh:87]: Nice catch!  I missed adding those two that were present as part of the CCv0 branch but not here.

**Pull request merged**